src/protocol/oprf/shuffle/mod.rs

```diff
@@ -0,0 +1,412 @@
+    }
+}
+
+async fn run_h1<C, I, S, Zl, Zr>(
+    ctx: &C,
+    batch_size: usize,
+    shares: I,
+    (z_31, z_12): (Zl, Zr),
+) -> Result<Vec<AdditiveShare<S>>, Error>
+where
+    C: Context,
+    I: IntoIterator<Item = AdditiveShare<S>>,
+    S: SharedValue + Add<Output = S> + Message,
+    Zl: IntoIterator<Item = S>,
+    Zr: IntoIterator<Item = S>,
+    for<'a> &'a S: Add<Output = S>,
+    Standard: Distribution<S>,
+{
+    // 1. Generate helper-specific random tables
+    let ctx_a_hat = ctx.narrow(&OPRFShuffleStep::GenerateAHat);
+    let a_hat = generate_random_table_solo(batch_size, &ctx_a_hat, Direction::Left);
+
+    let ctx_b_hat = ctx.narrow(&OPRFShuffleStep::GenerateBHat);
+    let b_hat = generate_random_table_solo(batch_size, &ctx_b_hat, Direction::Right);
+
+    // 2. Run computations
+    let a_add_b_iter = shares
+        .into_iter()
+        .map(|s: AdditiveShare<S>| s.left().add(s.right()));
+    let mut x_1: Vec<S> = add_single_shares(a_add_b_iter, z_12).collect();
+
+    let ctx_perm = ctx.narrow(&OPRFShuffleStep::ApplyPermutations);
+    let (mut rng_perm_l, mut rng_perm_r) = ctx_perm.prss_rng();
+    x_1.shuffle(&mut rng_perm_r);
+
+    let mut x_2 = x_1;
+    add_single_shares_in_place(&mut x_2, z_31);
+    x_2.shuffle(&mut rng_perm_l);
+    send_to_peer(&x_2, ctx, &OPRFShuffleStep::TransferX2, Direction::Right).await?;
+
+    let res = combine_single_shares(a_hat, b_hat).collect::<Vec<_>>();
+    Ok(res)
+}
+
+async fn run_h2<C, I, S, Zl, Zr>(
+    ctx: &C,
+    batch_size: usize,
+    shares: I,
+    (z_12, z_23): (Zl, Zr),
+) -> Result<Vec<AdditiveShare<S>>, Error>
+where
+    C: Context,
+    I: IntoIterator<Item = AdditiveShare<S>>,
+    S: SharedValue + Add<Output = S>,
+    Zl: IntoIterator<Item = S>,
+    Zr: IntoIterator<Item = S>,
+    for<'a> &'a S: Add<S, Output = S>,
+    for<'a> &'a S: Add<&'a S, Output = S>,
+    Standard: Distribution<S>,
+{
+    // 1. Generate helper-specific random tables
+    let ctx_b_hat = ctx.narrow(&OPRFShuffleStep::GenerateBHat);
+    let b_hat: Vec<S> =
+        generate_random_table_solo(batch_size, &ctx_b_hat, Direction::Left).collect();
+
+    // 2. Run computations
+    let c = shares.into_iter().map(|s| s.right());
+    let mut y_1: Vec<S> = add_single_shares(c, z_12).collect();
+
+    let ctx_perm = ctx.narrow(&OPRFShuffleStep::ApplyPermutations);
+    let (mut rng_perm_l, mut rng_perm_r) = ctx_perm.prss_rng();
+    y_1.shuffle(&mut rng_perm_l);
+
+    let mut x_2: Vec<S> = Vec::with_capacity(batch_size);
+    future::try_join(
+        send_to_peer(&y_1, ctx, &OPRFShuffleStep::TransferY1, Direction::Right),
+        receive_from_peer_into(
+            &mut x_2,
+            batch_size,
+            ctx,
+            &OPRFShuffleStep::TransferX2,
+            Direction::Left,
+        ),
+    )
+    .await?;
+
+    let mut x_3 = x_2;
+    add_single_shares_in_place(&mut x_3, z_23);
+    x_3.shuffle(&mut rng_perm_r);
+
+    let mut c_hat_1 = repurpose_allocation(y_1);
```

Comment:
This may be cheap but it has a non-trivial cost of zeroing out memory. I don't see a need to do that, you can just replace elements in that vector from the iterator.